File: Backend-KartingRM/reservas-comprobantes-service/src/main/java/com/kartingrm/reservas_comprobantes_service/service/ReservaService.java
package com.kartingrm.reservas_comprobantes_service.service;

import com.kartingrm.reservas_comprobantes_service.entity.Reserva;
import com.kartingrm.reservas_comprobantes_service.model.ClienteDTO;
import com.kartingrm.reservas_comprobantes_service.model.PlanDTO;
import com.kartingrm.reservas_comprobantes_service.repository.ReservaRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpEntity;
import org.springframework.stereotype.Service;
import org.springframework.web.client.RestTemplate;

import javax.persistence.EntityNotFoundException;
import java.time.LocalDate;
import java.time.LocalTime;
import java.util.List;
import java.util.Optional;

@Service
public class ReservaService {

    @Autowired
    private ReservaRepository reservaRepository;

    @Autowired
    RestTemplate restTemplate;


    public List<Reserva> getReservas() {
        return reservaRepository.findAll();
    }

    public Reserva getReservaById(Long id) {
        Optional<Reserva> reserva = reservaRepository.findById(id);
        if (reserva.isEmpty()) throw new EntityNotFoundException("Reserva " + id + " no encontrada");

        return reserva.get();
    }

    // Obtener todas las reservas de un cliente segun id
    public List<Reserva> getReservasByIdReservante(Long idCliente) {
        return reservaRepository.findReservasByIdReservante(idCliente);
    }


    // Crear reserva
    // Considera que no existan reservas previas en el horario, ademas ingresa automaticamente la hora de fin segun el plan
    // Verifica que sea feriado????


    // El cuerpo ya tiene id cliente reservante e id plan
    public Reserva createReserva(Reserva reserva) {
        // Obtener Cliente
        ClienteDTO cliente = restTemplate.getForObject("http://cliente-desc-frecu-service/api/cliente-service/cliente/" + reserva.getIdReservante(), ClienteDTO.class);
        // Obtener Plan
        PlanDTO plan = restTemplate.getForObject("http://plan-service/api/plan/planes/" + reserva.getIdPlan(), PlanDTO.class);

        // Verificar parametros validos
        if (reserva.getHoraInicio() != null && cliente != null && plan != null) {

            // Determinar hora de fin sumando minutos del plan a la hora inicial
            LocalTime horaInicio = reserva.getHoraInicio();
            LocalTime horaFinalCalculada = horaInicio.plusMinutes(plan.getDuracionTotal());
            reserva.setHoraFin(horaFinalCalculada);

            // Verificar que no existan reservas enter ambas horas
            if (existeReservaEntreDosHoras(reserva.getFecha(), horaInicio, horaFinalCalculada))
                throw new IllegalStateException("Ya existe una reserva con ese horario");



            // Verificar Horario de inicio y fin validos.
            //  Lunes a Viernes: 14:00 a 22:00
            //  Sabados, Domingos, Feriados: 10:00 a 22:00
            boolean esFinDeSemana = reserva.getFecha().getDayOfWeek().getValue() >= 6;// Determina fin de semana o no

            // Obtener bool si es feriado
            boolean esFeriado = restTemplate.getForObject("http://dias-especiales-service/api/dias-especiales-service/dias-feriados/esFeriado?fecha=" + reserva.getFecha(),Boolean.class);
            if (esFeriado || esFinDeSemana){// Horario fin de semana o feriado
                // Horario antes o despues del horario de servicio
                if (horaInicio.isBefore(LocalTime.of(10,00,00)) || horaFinalCalculada.isAfter(LocalTime.of(22,00,00))) {
                    throw new IllegalStateException("Horario incorrecto. Domingos, sábados y feriados: 10:00 a 22:00");
                }
            } else { // Horario semana
                if (horaInicio.isBefore(LocalTime.of(14,00,00)) || horaFinalCalculada.isAfter(LocalTime.of(22,00,00))) {
                    throw new IllegalStateException("Horario incorrecto. Lunes a Viernes: 14:00 a 22:00");
                }
            }


            return reservaRepository.save(reserva);
        } else {
            throw new RuntimeException("Cliente no encontrado, plan no existe, o hora errónea");
        }
    }


    // Update de valor de reserva. Permite cambio de plan y de cliente reservante
    public Reserva updateReserva(Long id, Reserva reserva) {
        Optional<Reserva> reservaOriginalOptional = reservaRepository.findById(id);
        if (reservaOriginalOptional.isEmpty()) throw new EntityNotFoundException("Reserva id " + id + " no encontrado");

        // Reserva actualizada conserva id de la reserva original
        reserva.setId(id);

        // Calculo de hora final segun nueva hora o plan
        // Obtener Plan
        PlanDTO plan = restTemplate.getForObject("http://plan-service/api/plan/planes/" + reserva.getIdPlan(), PlanDTO.class);

        // Solo permite actualizar la hora inicial
        // Determinar hora de fin sumando minutos del plan a la hora inicial
        LocalTime horaInicio = reserva.getHoraInicio();
        LocalTime horaFinalCalculada = horaInicio.plusMinutes(plan.getDuracionTotal());
        reserva.setHoraFin(horaFinalCalculada);

        // Verificar que no existan reservas enter ambas horas
        if (existeReservaEntreDosHoras(reserva.getFecha(), horaInicio, horaFinalCalculada))
            throw new IllegalStateException("Ya existe una reserva con ese horario");

        // Verificar horario valido de atencion
        boolean esFinDeSemana = reserva.getFecha().getDayOfWeek().getValue() >= 6;// Determina fin de semana o no
        boolean esFeriado = restTemplate.getForObject("http://dias-especiales-service/api/dias-especiales-service/dias-feriados/esFeriado?fecha=" + reserva.getFecha(),Boolean.class);
        if (esFeriado || esFinDeSemana){// Horario fin de semana o feriado
            // Horario antes o despues del horario de servicio
            if (horaInicio.isBefore(LocalTime.of(10,00,00)) || horaFinalCalculada.isAfter(LocalTime.of(22,00,00))) {
                throw new IllegalStateException("Horario incorrecto. Domingos, sábados y feriados: 10:00 a 22:00");
            }
        } else { // Horario semana
            if (horaInicio.isBefore(LocalTime.of(14,00,00)) || horaFinalCalculada.isAfter(LocalTime.of(22,00,00))) {
                throw new IllegalStateException("Horario incorrecto. Lunes a Viernes: 14:00 a 22:00");
            }
        }

        return reservaRepository.save(reserva);
    }


    public boolean deleteReserva(Long id) {
        Optional<Reserva> reservaOriginalOptional = reservaRepository.findById(id);
        if (reservaOriginalOptional.isEmpty()) throw new EntityNotFoundException("Reserva id " + id + " no encontrado");

        reservaRepository.deleteById(id);
        return true;
    }



    //  Metodos privados. Logica interna

    // Obtener reservas existentes entre dos horas de un dia
    // Se usa en como condicion para crear reservas
    private boolean existeReservaEntreDosHoras(LocalDate fecha, LocalTime horaInicio, LocalTime horaFinal) {
        return reservaRepository.existeReservaEntreDosHoras(fecha, horaInicio, horaFinal);
    }


    // Metodos para obtener info para reportes

}

File: Front KartingRM/src/services/reservaService.js
import httpClient from '../http-common';

const URL_LOCAL = '/reservas';

// Peticion GET de todas las reservas
export const getReservas = async () => {
  try {
    const response = await httpClient.get(`${URL_LOCAL}/`);
    return response.data;
  } catch (error) {
    console.error('Error al obtener reservas:', error);
    throw error;
  }
};

// Peticion Get de reserva segun id
export const getReservaById = async (idReserva) => {
  try {
    const response = await httpClient.get(`${URL_LOCAL}/${idReserva}`);
    return response.data;
  } catch (error) {
    console.error(`Error al obtener reserva con ID ${idReserva}:`, error);
    throw error;
  }
};

// Peticion Get de reservas de un cliente segun id
export const getReservasByCliente = async (idCliente) => {
  try {
    const response = await httpClient.get(`${URL_LOCAL}/por-cliente/${idCliente}`);
    return response.data;
  } catch (error) {
    console.error(`Error al obtener reservas del cliente ${idCliente}:`, error);
    throw error;
  }
};

// Peticion POST para crear reserva indicando id de cliente, id de plan, booleano que indica si 
//  es dia feriado, y el cuerpo de la reserva
export const createReserva = async (idCliente, idPlan, feriado, reservaData) => {
  try {
    const response = await httpClient.post(
      `${URL_LOCAL}?id_cliente=${idCliente}&id_plan=${idPlan}&feriado=${feriado}`,
      reservaData
    );
    return response.data;
  } catch (error) {
    console.error('Error al crear reservaxx:', error);
    throw error;
  }
};

// Peticion PUT para update de reserva segun cuerpo
export const updateReserva = async (idReserva, reservaData) => {
  try {
    const response = await httpClient.put(`${URL_LOCAL}/${idReserva}`, reservaData);
    return response.data;
  } catch (error) {
    console.error(`Error al actualizar reserva con ID ${idReserva}:`, error);
    throw error;
  }
};

// Peticion PUT para update del cliente reservante de una reserva
export const updateClienteReserva = async (idReserva, idCliente) => {
  try {
    const response = await httpClient.put(`${URL_LOCAL}/${idReserva}/cliente?id_cliente=${idCliente}`);
    return response.data;
  } catch (error) {
    console.error(`Error al actualizar cliente en reserva ${idReserva}:`, error);
    throw error;
  }
};

// Peticion PUT para update del plan de una reserva
export const updatePlanReserva = async (idReserva, idPlan) => {
  try {
    const response = await httpClient.put(`${URL_LOCAL}/${idReserva}/plan?id_plan=${idPlan}`);
    return response.data;
  } catch (error) {
    console.error(`Error al actualizar plan en reserva ${idReserva}:`, error);
    throw error;
  }
};

// Peticion PATCH para relacionar cliente integrante con reserva
export const addIntegrante = async (idReserva, idCliente) => {
  try {
    const response = await httpClient.patch(`${URL_LOCAL}/${idReserva}/integrantes/agregar/${idCliente}`);
    return response.data;
  } catch (error) {
    console.error(`Error al agregar integrante a reserva ${idReserva}:`, error);
    throw error;
  }
};

// Peticion PATCH para quitar un cliente como integrante de una reserva
export const removeIntegrante = async (idReserva, idCliente) => {
  try {
    const response = await httpClient.patch(`${URL_LOCAL}/${idReserva}/integrantes/quitar/${idCliente}`);
    return response.data;
  } catch (error) {
    console.error(`Error al quitar integrante de reserva ${idReserva}:`, error);
    throw error;
  }
};

// Peticion DELETE para eliminar una reserva
export const deleteReserva = async (idReserva) => {
  try {
    await httpClient.delete(`${URL_LOCAL}/${idReserva}`);
    return idReserva;
  } catch (error) {
    console.error(`Error al eliminar reserva con ID ${idReserva}:`, error);
    throw error;
  }
};